From 3962737e779dd744e95d175af6de6fa18c69edc0 Mon Sep 17 00:00:00 2001
From: Takashi Yoshii <takashi.yoshii.zj@renesas.com>
Date: Wed, 6 Mar 2013 18:03:09 +0900
Subject: [PATCH 14/17] mmc: sh-mmcif: support single irq variant

The IP on r8a73a4 handles all events by single irq.

Switch the behavior by the numbers of irq resource
1, handle all the events with it.
2, #0 is for errors, #1 is others. (backward compatible)

Signed-off-by: Takashi Yoshii <takashi.yoshii.zj@renesas.com>
---
 drivers/mmc/host/sh_mmcif.c |   25 +++++++++++++++++--------
 1 file changed, 17 insertions(+), 8 deletions(-)

diff --git a/drivers/mmc/host/sh_mmcif.c b/drivers/mmc/host/sh_mmcif.c
index 82bf921..9873c61 100644
--- a/drivers/mmc/host/sh_mmcif.c
+++ b/drivers/mmc/host/sh_mmcif.c
@@ -1307,11 +1307,15 @@ static int __devinit sh_mmcif_probe(struct platform_device *pdev)
 	char clk_name[8];
 
 	irq[0] = platform_get_irq(pdev, 0);
-	irq[1] = platform_get_irq(pdev, 1);
-	if (irq[0] < 0 || irq[1] < 0) {
+	if (irq[0] < 0) {
 		dev_err(&pdev->dev, "Get irq error\n");
 		return -ENXIO;
 	}
+	irq[1] = platform_get_irq(pdev, 1);
+	if (irq[1] < 0) {
+		irq[1] = irq[0];
+		irq[0] = 0;
+	}
 	res = platform_get_resource(pdev, IORESOURCE_MEM, 0);
 	if (!res) {
 		dev_err(&pdev->dev, "platform_get_resource error.\n");
@@ -1374,10 +1378,13 @@ static int __devinit sh_mmcif_probe(struct platform_device *pdev)
 	sh_mmcif_sync_reset(host);
 	sh_mmcif_writel(host->addr, MMCIF_CE_INT_MASK, MASK_ALL);
 
-	ret = request_threaded_irq(irq[0], sh_mmcif_intr, sh_mmcif_irqt, 0, "sh_mmc:error", host);
-	if (ret) {
-		dev_err(&pdev->dev, "request_irq error (sh_mmc:error)\n");
-		goto ereqirq0;
+	if (irq[0] > 0) {
+		ret = request_threaded_irq(irq[0], sh_mmcif_intr,
+				sh_mmcif_irqt, 0, "sh_mmc:error", host);
+		if (ret) {
+			dev_err(&pdev->dev, "request_irq error (sh_mmc:error)\n");
+			goto ereqirq0;
+		}
 	}
 	ret = request_threaded_irq(irq[1], sh_mmcif_intr, sh_mmcif_irqt, 0, "sh_mmc:int", host);
 	if (ret) {
@@ -1409,7 +1416,8 @@ emmcaddh:
 erqcd:
 	free_irq(irq[1], host);
 ereqirq1:
-	free_irq(irq[0], host);
+	if (irq[0] > 0)
+		free_irq(irq[0], host);
 ereqirq0:
 	pm_runtime_suspend(&pdev->dev);
 eresume:
@@ -1456,7 +1464,8 @@ static int __devexit sh_mmcif_remove(struct platform_device *pdev)
 	irq[1] = platform_get_irq(pdev, 1);
 
 	free_irq(irq[0], host);
-	free_irq(irq[1], host);
+	if (irq[1] > 0)
+		free_irq(irq[1], host);
 
 	platform_set_drvdata(pdev, NULL);
 
-- 
1.7.9.5

